components/layout_2020/display_list/mod.rs

```diff
@@ -450,6 +455,74 @@ impl Fragment {
             self.build_display_list_for_text_decoration(fragment, builder, &rect, &color);
         }
 
+        // Text Selection
+        if let Some(range) = &fragment.selection_range {
+            let selection_start = range
+                .start
+                .saturating_sub(fragment.character_start_offset)
+                .min(glyphs.len());
+            let selection_end = range
+                .end
+                .saturating_sub(fragment.character_start_offset)
+                .min(glyphs.len());
+
+            if selection_start < selection_end {
+                let selection_start = glyphs[selection_start].point;
```

Comment:
It's very wrong to index the glyphs by unicode characters. You are going to need to use harfbuzz to turn character offsets into glyph offsets. You should check the harfbuzz APIs as they may use utf-8 codepoint offsets in which case you do not need to call `chars()` so much.

Reply:
I'm not really familiar with harfbuzz, is there somewhere within the codebase that I can reference?